Update Condition guidance

evankanderson opened the pull request:
**Which issue(s) this PR fixes**:
Fixes #4471

I took a first stab at updating these conventions based on https://github.com/kubernetes-sigs/cluster-api/issues/1658#issuecomment-584794287 and https://github.com/kubernetes-sigs/service-apis/pull/47#issuecomment-580461766.

/assign @smarterclayton @liggitt 

I'm happy to recommend different top-level conditions type names. Note that I recommended `Succeeded` over `Complete` as a top-level condition, because `Complete` doesn't indicate whether the Job succeeded or failed.

youngnick commented:
When looking at this across a bunch of types, I wondered how we could possibly reconcile the history of Conditions in a way that would be machine-parseable easily. The best I could come up with was recommending that new Conditions add a `polarity` field (or whatever the name ends up being) that indicates if Status `True` or `False` is the healthy condition. If that is assumed to be `False` when not present, then a library could make a fair guess at what the Status means regardless of when it was created.

In no way intended to block this PR, which I fully support.

mrbobbytables commented:
/ok-to-test

evankanderson commented:
(I thought I'd sent an update from mobile while I was on vacation, but I either hallucinated it or technology ate it)

I'd like to defer the discussion of how to migrate existing controllers into the new pattern to a subsequent PR, possibly as a KEP or introduction of common infrastructure for managing Conditions. However, since you brought it up, I think I'd prefer to introduce a `severity` field which could be used for both `ErrorIfTrue` and `ErrorIfFalse` signalling, but also indicate that some Conditions provide information without signalling a fatal error.

danehans commented:
@evankanderson thank you for putting this PR together. https://github.com/kubernetes-sigs/service-apis/pull/104 is on hold until this merges. @smarterclayton @liggitt is anything needed to move this PR forward?

therealmitchconnors commented:
These arguments are compelling, but run opposite of previous guidance used for projects like kstatus.  Can we get sig-architecture clarification on whether healthy conditions should be true or false?

erictune commented:
Big fan of this change.

liggitt commented on contributors/devel/sig-architecture/api-conventions.md:
```diff
@@ -374,9 +394,9 @@ and should assume an Open World.
 An example of an oscillating condition type is `Ready` (despite it running
 afoul of current guidance), which indicates the object was believed to be fully
```
if this changes the guidance around positive/negative conditions, we can drop the `(despite it running afoul of current guidance)` parenthetical

evankanderson replied:
Thanks for catching this, updated.

youngnick commented:
@therealmitchconnors This PR is codifying the answers you'll get from people if you ask around into the official docs, so we can all be on the same page from now on. Any stuff that's already done will stay how it is, but when I asked sig-arch and sig-api-machinery, I was told this was the new way. I agree that having someone confirm that here would be ideal though.

liggitt commented:
cc @deads2k re https://github.com/kubernetes/enhancements/pull/1624

liggitt commented on contributors/devel/sig-architecture/api-conventions.md:
```diff
@@ -308,17 +308,47 @@ response reduces the complexity of these clients.
+  changed arbitrarily - it becomes part of the API, and has the same backwards-
+  and forwards-compatibility concerns of any other part of the API.
+
+* Conditions should have a consistent polarity. **A "positive" polarity where
+  "True" indicates normal operation and "False" indicates a failure is
+  recommended.** (Note that this is an explicit reversal from earlier
```
Clarify this is a guide for choosing condition names and polarity, not a guide for consuming arbitrary conditions. As written, it sounds like I can write a client that assumes all conditions with a `False` status represent failure.

evankanderson replied:
I would like to reach that point sometime in the future, but you're correct that we can't do this for a long time, if ever.

Adjusted this to suggest that consistency within a resource was most important consideration, and positive polarity should be chosen if there isn't already a consistent polarity.

liggitt commented on contributors/devel/sig-architecture/api-conventions.md:
```diff
@@ -308,17 +308,47 @@ response reduces the complexity of these clients.
+     may require adjusting condition type names, for example changing the type
+     name from "QuotaExhausted" to "ResourcesAvailable".
+
+   * The `Unknown` status is the default, and typically indicates that
```
The term `default` is overloaded... this makes it sound as though a condition that omits the status field has its status defaulted to `Unknown`, which is not the case. The phrase later is clearer: 

> The absence of a condition should be interpreted the same as `Unknown`.

evankanderson replied:
Done.

liggitt commented on contributors/devel/sig-architecture/api-conventions.md:
```diff
@@ -308,17 +308,47 @@ response reduces the complexity of these clients.
+     reconciliation has not yet finished (or that the resource state may not
+     yet be observable).
+
+* It's helpful to have a common top-level condition which summarizes more
+  detailed conditions. Simple consumers may simply query the top-level
+  condition. The `Ready` and `Succeeded` condition types may be used for
+  long-running and bounded-execution objects, respectively.
```
this also needs to be clearly oriented at the condition producer; as phrased it could be misunderstood by a simple consumer to mean that they can consistently expect to be able to use `Ready` / `Succeeded` conditions.

evankanderson replied:
Again, this is the long-term situation that I'd like to steer the world towards, but you're correct that this could be read by many consumers to demand more from API implementers than is likely in the short term.

Amended, PTAL.

liggitt commented:
I think this is a good improvement and is at a reasonable place to merge.

cc @kubernetes/api-reviewers for visibility, I plan to ack end of week

thockin commented on contributors/devel/sig-architecture/api-conventions.md:
```diff
@@ -308,17 +308,49 @@ response reduces the complexity of these clients.
+controllers to collect summary information about resources without needing to
+understand resource-specific status details. Conditions should complement more
+detailed information about the observed status of an object written by a
+controller, rather than replace it. For example, the "Available" condition of a
```
So this really inverts the previous guidance, which was sort of agonized over (and with which I happen to agree). Available (or Ready in the previous edit) are omnipresent and (I think?) are never "unknown".

Why are those correct as Conditions rather than fields?  What principal guides one to decide between those?

thockin replied:
If I have only one argument to make - it's this one.  As written, many status fields should just become conditions, and I don't think that's right.

thockin replied:
If we were going to have a distinct subresource for conditions, which we could RBAC distinctly, then I could see an argument forming around it.  That doesn't answer the question for "core" conditions like ready.

lavalamp replied:
I agree with Tim.

lavalamp replied:
I'd need to look at those more carefully before I could agree that Deployment's conditions are a good example.

thockin replied:
None of this speaks to the question of principle - what is the litmus test that helps me decide - field or condition?

evankanderson replied:
I'd propose the following guidance:

1. Expose the underlying measurements used to make decisions as fields.
1. If the outcome of that decision is interesting to a human (roughly: does it affect the behavior or functionality of the object in an externally observable way), provide a _single_ Condition which summarizes whether that aspect of the controller algorithm is operating correctly ("within the envelope").

smarterclayton replied:
Conditions summarize characteristics of object lifecycle that must be generic across multiple types, have a need to convey both state and human message, and can be extended by third parties to add more nuance to existing lifecycle.  

An object in Kube is a distributed state machine.  The state specified to that object should always be fields.  Some objects have consistent fields that describe status of the state machine.  One of those is observedGeneration.  Another one is conditions.  Some conditions define generic, multi-object state machine characteristics like Available, Progressing, Ready.

thockin commented on contributors/devel/sig-architecture/api-conventions.md:
```diff
@@ -308,17 +308,49 @@ response reduces the complexity of these clients.
+
+* Conditions should be added to explicitly convey properties that users and
+  components care about rather than requiring those properties to be inferred
+  from other observations.  Once defined, the meaning of a Condition can not be
```
Maybe worth expanding on "meaning"?  Specifically roll-up sorts of signals are valuable because consumers do not have to do look at N raw signals and make the same inference AND because that set of N raw signals can evolve into M raw signals, and the roll-up is handled automatically.

See for example pod readiness gates - the raw signals got more complex, but consumers only had to worry about the "Ready" condition.

In that case the internal composition of "Ready" changed, but the external semantic did not.

lavalamp replied:
We can't really enforce this in code, so users can't really rely on this. I'd therefore phrase it as a recommendation rather than requirement.

evankanderson replied:
I think "meaning" here is the "intent" of the Condition -- it may become possible to more accurately measure the intent.

For example, adding "is there enough disk space" to the set of conditions that indicate "this Node is able to accept work" or "the Pods on this Node are likely to function properly" -- but it would not be acceptable to swap "able to accept work" intent for an "Pods function properly" intent; the latter would need to be a new Condition.

thockin replied:
My original comment was that the text could be clearer about what "meaning" means - how abstract that is vs how it is literally derived.

But I think it also unclear if you mean that conditions should exclusively be used as "summaries" of other information or that "summaries" are one of several reasonable uses of conditions?

And I am still looking for answers to my question on principle.  Why is a "ready" pronouncement *better* a a condition than as a field?

seaneagan replied:
My read is that conditions have consistent metadata attached to them (at least as of #1624) as to when and why they last transitioned states, including now the `observedGeneration`, which was previously not specific to conditions, so I guess that could be considered a new advantage of conditions over fields, or alternatively a new disadvantage of fields as it may make non-condition metadata harder to reason about. I suppose it doesn't stop anyone from defining `status.observedGeneration` as well, but overall having both seems like potential API bloat.

thockin commented on contributors/devel/sig-architecture/api-conventions.md:
```diff
@@ -308,17 +308,49 @@ response reduces the complexity of these clients.
+  changed arbitrarily - it becomes part of the API, and has the same backwards-
+  and forwards-compatibility concerns of any other part of the API.
+
+* Within a resource, Conditions should have a consistent polarity. **A
```
I don't think we can say this *and* say that conditions are an extension mechanism.

evankanderson replied:
Unless we define the polarity for extension mechanism... (and I realize that Pod readiness conditions are currently mixed polarity, which complicates this).

lavalamp replied:
The point is that we can't enforce polarity with validation code, and there's no limit on who can make conditions. There's no top-down design for the set of conditions on a particular object. They can come from different sources. (Node is a good example of this.)

thockin commented on contributors/devel/sig-architecture/api-conventions.md:
```diff
@@ -308,17 +308,49 @@ response reduces the complexity of these clients.
+* Within a resource, Conditions should have a consistent polarity. **A
+  "positive" polarity where "True" indicates normal operation and "False"
+  indicates a failure is recommended if there is not an established polarity.**
+  (Note that this is an explicit reversal from earlier recommendations of
```
Have we surveyed all of the built-in conditions for this?

Just at a glance:
* Node has Ready (positive) and *Pressure and NetworkUnavailable (all negative)
* Namespace has *Failure (negative) and *Remaining (ambiguous)
* PVC has Resizing (negative)
* Pod has all positive
* RC has failure
* Deployment, HPA are mixed

evankanderson replied:
Looking through GoDoc at https://godoc.org/k8s.io/api :
* At least 12 packages have nothing exported matching "ConditionType"

* HPA's "ScalingLimited" is not clearly a failure, but might be an exceptional case.
* Job has "Complete" and "Failed" (and "Succeeded" is *Complete=True and Failed unset*)
* ReplicaSet has only "ReplicaFailure" (mirroring ReplicationController, and also mirrored into Deployment), and DaemonSet and StatefulSet have no actual Condition types, despite defining types (!)
* PriorityLevel and FlowSchema also define a type but no conditions
* PVC's "Resizing" isn't positive/negative -- it's a non-disruptive indication of work in progress (and failure to expand the volume isn't indicated by Condition at all)

I think it's fair to characterize this as "all over the place".

Unfortunately, my experience is that lacking a consistent "is this working" / "what's not working" mechanism makes Kubernetes less approachable for newcomers, who end up needing to cram the entire system into their head when debugging (or throw up their hands and find someone who has already done so).

lavalamp replied:
I agree that's a defect, but I don't think better defining conditions is going to solve it.

I think it'd be fun to write a "quiescence controller" that watches existing objects, and maintains a "quiescent" condition. Then we'd just have to write that logic in one place.

thockin replied:
If we want to say that there's a convention on polarity, we should commit to going through EVERY SINGLE built-in condition, deprecating the ones that are "wrong" (can't delete, but...) and adding an equivalent-opposite in the "correct" polarity.

smarterclayton replied:
I don't like the polarity convention, except that conditions must summarize to both humans and machines, and the polarity should make sense to a human in the context of the condition type, not as part of some aribrarty rule.

NotReady = true is strictly harder for a human to understand than Ready = false and for that reason alone should be excluded.  We choose to name fields by what they do, rather than by how they are introduced - why would we have different rules for conditions?

lavalamp replied:
I agree w/ Clayton; we probably got into this situation by not having a clear understanding that a condition not being present means "unknown" not "false".

Unfortunately, "polarity is for people" means that a aggregating controller like I talk about above needs an extra bit of information, but I think that's still tractable.

erictune replied:
This seems like the most contentious part of this PR.   Would the author like to replace the proposed changes to "polarity" with something like:

````For some conditions, True represents normal operation, and for some "False" represents normal operation.  It is not possible to compute an overall "summary" condition from all the conditions in an object in a a generic way.```

`without detailed knowledge of the specific  are Tpull out the polarity changes, or maybe just change the existing language about polarity to acknowledge that

lavalamp commented:
@deads2k is trying to regularize these and should probably look at this. I intend to look at this soon.

thockin commented:
/assign

liggitt commented:
ack. will let discussion settle and coordinate with @deads2k's PR

lavalamp commented on contributors/devel/sig-architecture/api-conventions.md:
```diff
@@ -371,11 +393,11 @@ we define comprehensive state machines for objects, nor behaviors associated
+object was believed to be fully operational at the time it was last probed. A
+possible monotonic condition could be `Succeeded`. A `True` status for
+`Succeeded` would imply completion and that the resource was no longer
+active. An object that was still active would generally not have a `Succeeded`
```
Note that `Failed` and `Succeeded` are not actually opposites and wouldn't communicate the same information. If you switched this because of the "prefer positive polarity" guidance, it's maybe a good example of something that needs an exception.

evankanderson replied:
Agree that "Failed" and "Succeeded" aren't opposites. Job is actually a great example of this -- it has two Conditions ("Complete" and "Failed"). The possible states for a Job are:

| Complete | Failed | meaning |
| --- | --- | --- |
| unset | unset | Still running |
| True | unset | Succeeded |
| True | True | Failed |

Note that many (most?) Kubernetes objects indicate a desire for an ongoing condition, rather than a one-time run, so while "Failed" may indicate having come to a flaming halt, but the opposite is probably "hasn't come to a flaming halt _yet_", i.e. "Ready", rather than "Succeeded".

lavalamp commented on contributors/devel/sig-architecture/api-conventions.md:
```diff
@@ -308,17 +308,49 @@ response reduces the complexity of these clients.
+structure. Conditions are most useful when they follow some consistent
+conventions:
+
+* Conditions should be added to explicitly convey properties that users and
```
I don't really agree. I think there's room in the world for two kinds of conditions:
* "Reportables"-- conditions that may or may not be true of particular subsets of items. There's many ways of reporting these and we don't expect consumers to understand them.
* "Consumables"-- these conditions are summaries of the other facts about the object, packaged for consumers that need a high-level bit to trigger off of.

This assumes some sort of bridge controller which understands all the reportables and can summarize them into the consumables.

evankanderson replied:
If we allow Condition types to be namespaced (Tim's idea), it would be possible to summarize all namespaced conditions with a consistent polarity using a "bridge controller".

Existing Conditions would need some sort of annotation to indicate their expected polarity, since they are generally mixed.

lavalamp replied:
I think there's two ideas here?
* namespacing is usually just to avoid naming collisions. I'm pretty sure we decided against namespacing these way back when specifically to force people to settle on Schelling point names & meanings. I'm not sure if this worked or not. It's definitely going to be confusing to have different conditions with the same name but different namespaces.
* "aspect of": I think many conditions have this relation, where e.g. network reachability is an "aspect of" a pod's overall readiness. If we codified this explicitly, it'd be easy to have a controller aggregate all aspects of a given top-level consumable. This wouldn't necessarily require consistent polarity with the top level consumable (although that'd be most obvious), you could represent with with a signed value, where e.g. -1/0/+1 means false-disables / no effect / true-disables. This could get arbitrarily complicated, though, and I think it might be better to just express it in code in a controller.

thockin replied:
I'm lost on this thread - I don't know what a "bridge controller" is for?

Regarding polarity, we could have an indicator, as Daniel suggests, so it is self-describing.  But "good/bad" is a pretty vague proclamation.  And we'd have to make it optional ANYWAY, so ...

If the goal is to have machines understand arbitrary conditions, we need an indicator of normal or abnormal polarity.  Is that the goal?  Or do we expect that most consumers of conditions are either a) humans who can read the name and infer polarity; b) programs coded by humans to look for specific conditions with a priori known polarity?

We can't simply say that Conditions are "summarizations" of other info, because they are ALSO the only extension point for external status.

evankanderson replied:
I'd like to reach the point where machines can manage arbitrary Conditions. Evolving that from the current state may require additional indicator fields, depending on implementation strategy. If there are existing status contributors with mixed polarity within an object, we'll definitely need a "Polarity" field. If we perform the status aggregation in the "main" controller for the object, we can do this object-by-object and might be able to avoid Polarity.

I interpreted @lavalamp 's idea of a "bridge controller" as a single controller which could operate across all Kinds and perform this aggregation based on mechanical rules. Doing this aggregation within the object's status avoids duplicating the aggregation logic across many clients; doing this in a "bridge controller" avoids duplicating the aggregation in many controllers (which may or may not be as much of an issue).

I'm not sure how many aggregated conditions are needed (vs creating additional objects for "consumable" measurements that are composite). We got away with one in Knative, but singletons always come back to greet you. The advantage of codifying the condition aggregation of "aspect of" in a controller is that it allows new controllers to contribute observations without needing to modify the logic of the main controller to recognize the new conditions.

Playing with an experiment we did in Knative, it would be possible to add "Severity" and "Polarity" fields:

| Severity | Polarity | Effects |
|---|---|---|
| Error (default / empty-string) | Positive | If "False", "Ready" is False |
| Error (default / empty-string) | Negative | If "True", "Ready is False |
| Warning | Positive | No effect on "Ready", but may be highlighted by UIs |
| Warning | Negative | No effect on "Ready", but may be highlighted by UIs, e.g. "ReplicaFailure" in Deployment |
| Info | N/A | Informational observations, such as "Resizing" in PVC |
| unset | unset | Legacy Condition, not aggregated 🗡️ |

🗡️ - For legacy conditions, we might want to prioritize updating those that should be treated as "Error" to set a polarity, then those that should be prioritized as "Warning".

If we proceed object-at-a-time, the Polarity field might be able to be dropped, but this would probably require a certain amount of explicit handling of legacy conditions.

thockin replied:
I'm not inherently against more self-describing, extensible status, but  tat's a significant change from Conditions today.

Ready is a condition of its own and we don't have "virtual" conditions.  So clients would need to be taught this new structure, which is obviously backwards-incompatible.  CF the design around Pod `readinessGates` which had a constraint that clients who only know to read `Ready` would still be correct.

thockin replied:
In other words, if you want something like this, I think it needs to be totally additive (a new concept)

bowei commented:
/cc @bowei

Jamstah commented:
Coming here from https://github.com/kubernetes/enhancements/pull/1624 because I think the comment might be more useful here:

Do we want to present an opinion on the ordering of the conditions array in this proposal?

For consumers, it would be good to know if this array should be presented in the order it is given (which I expect is the right approach), so that controllers managing the status array can provide whatever significant ordering they feel is appropriate, and have that reflected in tooling for human users.

liggitt commented:
I'd expect the conditions list to be a listType=map, which means it is keyed by type and the server (in some cases) will manage merging added conditions. In lists like that, order is not typically significant.

evankanderson commented:
+1 to order not being significant.

I think there are at least three items here:

1. Do we want to reverse the polarity recommendation? Early feedback from @smarterclayton and @liggitt suggested that we might, but @thockin seems skeptical. My _personal_ experience is that the current recommendation (negative polarity) led to a higher number of "working as intended" issues when people read "error name" in the status block and didn't read "False" two lines down.

1. Do we want to recommend using Conditions as a high-level status mechanism? If so, it's probably worth providing additional guidance on how to use them (and, if we think the Node `status` contribution from multiple controllers is a good pattern, guidance on how multiple controllers should cooperate on Conditions). Personally, I'm a fan of using Conditions in this way, as it makes it easier to build generic UIs across CRDs -- everything with a "Ready=False" condition can be flagged with one icon, and everything with "Ready=Unknown" can be flagged with a different icon.

1. Do we want to provide a library to help with this? https://github.com/kubernetes/enhancements/issues/1623 provides a start to this, but standardizes the structure without standardizing the semantics. I'd be willing to collect a set of contributors to propose a KEP for this if we think that format would be useful.

dims commented:
/priority critical-urgent

dims commented:
email thread : https://groups.google.com/d/msg/kubernetes-sig-architecture/Es2FD-U5who/SbP-F8wcBgAJ

lavalamp commented:
Repeating my thoughts from an email thread on the SIG arch mailing list:

It's slightly silly to (effectively) repeat the schema for each condition in each instance of the condition, but the alternative is some sort of schema registry system which would add a lot of complexity. I don't think it's worth it right now, maybe in v2.

I'd recommend the policy:
* "Polarity is for people" - state the condition in a natural way for humans to understand.
* Absent means we don't know the value of the condition, and we *also* don't know if the setter of the condition is running/active.
* Setters of conditions should set the condition for the objects they're attempting to monitor, even if "Unknown" is the best they can do for a value at the moment. Otherwise, it's difficult to reason about the set of conditions that might suddenly show up on an object and change your opinion about it.
* [optional, up for debate] Add a `polarity: [AbnormalTrue | AbnormalFalse | InformationalOnly]` field. This is optional, we could also require consumers of the condition to understand by convention, as it is today.
* [optional, up for debate] Add a `modifies: [NameOfOtherCondition1, NameOfOtherCondition2]` field. This, along with polarity, allows consumers to aggregate conditions that they don't understand. It's basically the fully general form of pod readiness gates.

(I think (hope?) we can all agree on the first three; so below I'll discuss the last two.)

Downsides:
* Adding fields makes conditions more verbose
* Exposing the dependency graph causes some problems:
  * it either pushes additional work onto clients
  * or we need a condition aggregating controller
  * Objects will appear inconsistent at first glance (like `[Ready: True], [NetworkAttached: False, polarity: AbnormalFalse, modifies: Ready]` really means "ready: False")
  * We could validate that parent conditions have the value implied by their children, but then we'd lose the input of whatever was setting the parent condition.
  * Clients might ignore the modifier condition. We could solve that by making apiserver (or a mutating webhook) generate a modified judgement field in the parent condition, but clients might ignore that, too.

An alternative is to express the "condition schema" implied by my "polarity/modifies" fields in some other way.
* This adds another moving piece; it'd actually be more work for clients to assemble the information themselves.
* Suppose we implemented this with e.g. a CRD expressing the per-object-type condition relationships and a mutating webhook: then we'd still need to add a "modified judgement" field
* ...or we'd need to fabricate additional conditions ("ModifiedReady"); either way, clients would STILL have to change to get the full value.
* This doesn't solve the problem of clients ignoring things.

youngnick commented:
I agree that Conditions should first and foremost be for people, and the polarity should be easily inferrable by a person from the Condition's `Name`.

However, not including something to make the polarity machine-parseable ensures that anyone writing tooling to consume conditions will need to have a person see each possible condition at least once, so that a person can confirm the polarity. Without that bit of information, it's not possible to determine an arbitrary condition's meaning mechanically, and by not including something to do that, we're kind of implying that mechanically parsing Conditions is not desirable.

Obviously conventions like  the one proposed by [kstatus](https://github.com/kubernetes-sigs/cli-utils/tree/master/pkg/kstatus) are helpful, but if you want to write more general code, like say a UI or general monitoring library, you are out of luck. That's totally a decision we as a project can make, but it does come with an opportunity cost.

If we were to add a `polarity` field or similar, an option to reduce the amount of work to bring everything together might be to specify the new fields missing behavior as equivalent to `AbnormalTrue` - so if a Condition does not specify a polarity, it aligns with the current guidance.

This would create a reasonably straightforward set of rules for parsing arbitrary Conditions:
- Condition not present: `Unknown`, relevant controller hasn't touched the object yet. In this case, consumers of the Condition should backoff and try again later.
- Condition present, no polarity field: Condition is `AbnormalTrue`
- Condtion present, polarity field set: Condition is whatever the field is set to, `AbnormalTrue`, `AbnormalFalse`, `InformationalOnly`.

evankanderson commented:
> This would create a reasonably straightforward set of rules for parsing arbitrary Conditions:
> 
> * Condition not present: `Unknown`, relevant controller hasn't touched the object yet. In this case, consumers of the Condition should backoff and try again later.

Note that we need the "set condition as soon as object is seen" guidance so that third parties can know that the Condition exists / is possible to happen.

> * Condition present, no polarity field: Condition is `AbnormalTrue`
> * Condtion present, polarity field set: Condition is whatever the field is set to, `AbnormalTrue`, `AbnormalFalse`, `InformationalOnly`.

I don't think we can make this inference without putting a great delay into the tooling. Consider the case of using a kubectl 1.20 against a Kubernetes 1.17 cluster -- `kubectl get deploy foo` would return:

```yaml
status:
  conditions:
  - type: Available
    status: "True"
  - type: Progressing
    status: "True"
```

To the tool, this Deployment would appear to have an issue. Looking at the survey in https://github.com/kubernetes/community/pull/4521#discussion_r413100273, it looks like there's no good assumption (despite the API guidance) about which way conditions should point:

**Positive**
* Pod

**Negative**
* ReplicaSet
* Namespace?

**Mixed / Other**
* Node
* Deployment
* HPA?
* Job
* PVC

jpeach commented:
> Obviously conventions like the one proposed by [kstatus](https://github.com/kubernetes-sigs/cli-utils/tree/master/pkg/kstatus) are helpful, but if you want to write more general code, like say a UI or general monitoring library, you are out of luck. That's totally a decision we as a project can make, but it does come with an opportunity cost.

IIUC kstatus merely follows the existing guidance.
 
> If we were to add a `polarity` field or similar, an option to reduce the amount of work to bring everything together might be to specify the new fields missing behavior as equivalent to `AbnormalTrue` - so if a Condition does not specify a polarity, it aligns with the current guidance.
> 
> This would create a reasonably straightforward set of rules for parsing arbitrary Conditions:
> 
> * Condition not present: `Unknown`, relevant controller hasn't touched the object yet. In this case, consumers of the Condition should backoff and try again later.

How can you know whether a Condition is not present unless you know the full set of 
Conditions for the resource? Neither humans nor tools can know whether a Condition is not present.

evankanderson commented:
Thanks for the data, Eric!

The total score across the data set is 997, and there are 384 conditions, so the mean score for a condition name is about 3.

The total score for lowercase Conditions is 24. Interestingly, 7 of these are actually namespaced and dns-cased.

One thing that stands out to me (beyond the variation in conditions) is that only about 1/10th of the conditions use present tense verbs for naming ("Running", "Progressing", "Resizing", etc). Of the top 20 (which includes eveything with a score of 9 or above), only 2 use present-tense verbs ("Running" and "Progressing"). In comparison, adjectives like "Ready", "Complete", and "Available" have 10 uses, and past-tense verbs ("Failed", "Deployed", "Succeeded") have the remainder.

In the tail, we see more use of present-tense verbs, as well as a pattern of prefixing a common condition with something specific to the resource, like "QuayEcosystemProvisioningSuccess" or "InfraProvisioningFailed". The 260 composite word conditions (using a naive "exclude the last capital" algorithm) have a total score of 482, suggesting that prefixed names are somewhat less popular than average (unsurprisingly), and lending some additional weight to the notion that "Condition types are for humans". Popular multi-word conditions look like "ReplicaFailure", "DiskPressure", or "PodScheduled", where the qualifier is short and possibly disambiguates from other conditions on the same resource.

evankanderson commented:
Looking at the first 50 condition types by hand (total score 544):

27 are positive-polarity (total score 324)
13 are negative-polarity (total score 149)
10 are neutral ("Complete", "Provisioning", etc) (total score 71)

evankanderson commented:
Thanks for the suggested text, Eric!

I've removed the polarity recommendation, added some examples based on the patterns I extracted above, and added the recommendation to consistently set Conditions when controlling a resource, even if the status is "Unknown".

thockin commented:
I don't object to a polarity indicator at this point, but only because it's too                                                           
late.  I think it demonstrates a lack of structure.  When you say "I have a
condition" you don't generally interpret that as "normal".

@smarterclayton:

> NotReady = true is strictly harder for a human to understand than Ready = false and for that reason alone should be excluded.

I still hold that "ready" is exactly the sort of thing that should NOT be a
condition.  I also believe that MOST conditions (especially those that deserve
to be conditions) could be expressed in the "correct" polarity.

@lavalamp:

> Exposing the dependency graph ...  Objects will appear inconsistent

Yeah, what does this mean:

```
[Ready: True]
[NetworkAttached: False, polarity: AbnormalFalse, modifies: Ready]
[ImageHasCVEs: true, polarity: AbnormalTrue, modifies Ready]
```

We can define our way out of it, but ick. And if that chaining goes deeper.
It's human hostile, and so it is incompatible with the idea that conditions are
primarily for humans.

If we want to introduce dependent-conditions, maybe do it structurally.
There's your fully general readiness gates:

```
betterConditions:
   - name: Ready
     type: Collection,
     value: True
     subs:
       - name: WaitingForNetwork
         type: Value
         value: false
       - name: example.com/WaitingForLB
         type: Value
         value: true
       - name: ImageHasCVEs
         type: Value
         value: true
```
_________

@evankanderson:

> Note that we need the "set condition as soon as object is seen" guidance so that third parties can know that the Condition exists / is possible to happen.

If there is a risk of someone doing the wrong thing, this isn't sufficient.
The reason we have readinessGates the way they are is to close this race.

evankanderson commented:
> @evankanderson:
> 
> > Note that we need the "set condition as soon as object is seen" guidance so that third parties can know that the Condition exists / is possible to happen.
> 
> If there is a risk of someone doing the wrong thing, this isn't sufficient.
> The reason we have readinessGates the way they are is to close this race.

If the goal is to for controllers to be able to aggregate conditions in an extensible way to provide further actuation, I 100% agree that something like readinessGates is to avoid missing a condition. Unfortunately, adding a list of expected conditions ends up entangling the object`s `status` with knowledge of the controllers running in the cluster, which is appropriate for a data-plane requirement like Pod selection by a Service.

I *think* there's a second case for knowing the set of Conditions, which is for programmatic consumption by UIs (i.e. eventual human consumption, mediated by a machine). The second case is more along the lines of consistent status reporting/accounting than data-plane management.

I think that's what [lavalamp is referring to here](https://github.com/kubernetes/community/pull/4521#issuecomment-648942061):
> * Setters of conditions should set the condition for the objects they're attempting to monitor, even if "Unknown" is the best they can do for a value at the moment. Otherwise, it's difficult to reason about the set of conditions that might suddenly show up on an object and change your opinion about it.

[jpeach also brings up a similar concern](https://github.com/kubernetes/community/pull/4521#issuecomment-649865713).

Others' experience may differ; my experience working with UX designers was that users really *didn't* like it when the set of columns/colored dots representing conditions grew and shrank as the object was actuated by various controllers. (This assumes a UX view across a single kind where all objects in the collection are managed by the same controller set, which is a common but not universal case.)


Is it worth adding some advice/mention of the readinessGates mechanism if Conditions are being used to actuate the data plane?

thockin commented:
Evan: I can't tell if we are agreeing or not.  :P

> it's difficult to reason about the set of conditions that might suddenly
show up on an object and change your opinion about i

and

> users really didn't like it when the set of columns/colored dots
representing conditions grew and shrank as the object was actuated by
various controllers

...make me think you're agreeing, though I think my statement is more
exploratory than yours.  If your goal is to know every possible condition
on a resource, then ad hoc conditions are not sufficient.  I'm not sure
that knowing every possible condition is really the goal, though, is it?
For something like readiness it clearly matters that we don't have
false-positives, but readinessGates are not a "pleasant" API to use (mostly
because it is very localized).  If we broaden this pattern, we'll need
something easier to get right and harder to get wrong.


On Sun, Jun 28, 2020 at 8:14 AM Evan Anderson <notifications@github.com>
wrote:

> @evankanderson <https://github.com/evankanderson>:
>
> Note that we need the "set condition as soon as object is seen" guidance
> so that third parties can know that the Condition exists / is possible to
> happen.
>
> If there is a risk of someone doing the wrong thing, this isn't sufficient.
> The reason we have readinessGates the way they are is to close this race.
>
> If the goal is to for controllers to be able to aggregate conditions in an
> extensible way to provide further actuation, I 100% agree that something
> like readinessGates is to avoid missing a condition. Unfortunately, adding
> a list of expected conditions ends up entangling the objects status` with
> knowledge of the controllers running in the cluster, which is appropriate
> for a data-plane requirement like Pod selection by a Service.
>
> I *think* there's a second case for knowing the set of Conditions, which
> is for programmatic consumption by UIs (i.e. eventual human consumption,
> mediated by a machine). The second case is more along the lines of
> consistent status reporting/accounting than data-plane management.
>
> I think that's what lavalamp is referring to here
> <https://github.com/kubernetes/community/pull/4521#issuecomment-648942061>
> :
>
>
>    - Setters of conditions should set the condition for the objects
>    they're attempting to monitor, even if "Unknown" is the best they can do
>    for a value at the moment. Otherwise, it's difficult to reason about the
>    set of conditions that might suddenly show up on an object and change your
>    opinion about it.
>
> jpeach also brings up a similar concern
> <https://github.com/kubernetes/community/pull/4521#issuecomment-649865713>
> .
>
> Others' experience may differ; my experience working with UX designers was
> that users really *didn't* like it when the set of columns/colored dots
> representing conditions grew and shrank as the object was actuated by
> various controllers. (This assumes a UX view across a single kind where all
> objects in the collection are managed by the same controller set, which is
> a common but not universal case.)
>
> Is it worth adding some advice/mention of the readinessGates mechanism if
> Conditions are being used to actuate the data plane?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/4521#issuecomment-650777977>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVERVBXXYBSAQMVXTODRY5M6XANCNFSM4KVRDSEQ>
> .
>

jpeach commented:
While we are at it, TIL that condition types are expected to be [unique](https://github.com/kubernetes/kubernetes/pull/92519). 
The current API conventions don't address that, so I had assumed that using the same condition type with a different reason would be OK. When I think about conditions as error conditions, this seems overly restrictive to me, but when I think about them as state or phase markers, uniqueness seems desirable.

🤷

thockin commented:
What does "unique" mean here?

On Thu, Jul 2, 2020 at 2:56 PM James Peach <notifications@github.com> wrote:

> While we are at it, TIL that condition types are expected to be unique
> <https://github.com/kubernetes/kubernetes/pull/92519>.
> The current API conventions don't address that, so I had assumed that
> using the same condition type with a different reason would be OK. When I
> think about conditions as error conditions, this seems overly restrictive
> to me, but when I think about them as state or phase markers, uniqueness
> seems desirable.
>
> 🤷
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/4521#issuecomment-653236242>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVE32OEC7WYMO5VXOE3RZT67JANCNFSM4KVRDSEQ>
> .
>

jpeach commented:
> What does "unique" mean here?
> […](#)

IIUC, in any given slice of conditions, there should not be more that one condition with the same value for the `Type` field. All the condition slices in Kubernetes APIs are supposed to have `patchStrategy:"merge" patchMergeKey:"type"` annotations.

thockin commented:
oh, yeah.  I thought that was understood :)  Maybe I have just bumped my
head on non-unique key fields too many times.

On Thu, Jul 2, 2020 at 3:57 PM James Peach <notifications@github.com> wrote:

> What does "unique" mean here?
> … <#m_-7654946521204386007_>
>
> IIUC, in any given slice of conditions, there should not be more that one
> condition with the same value for the Type field. All the condition
> slices in Kubernetes APIs are supposed to have patchStrategy:"merge"
> patchMergeKey:"type" annotations.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/4521#issuecomment-653252502>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVBIFGG7XHASCFTCLNLRZUGFVANCNFSM4KVRDSEQ>
> .
>

jpeach commented:
> oh, yeah. I thought that was understood :) Maybe I have just bumped my head on non-unique key fields too many times.
> […](#)

I was proposing for a CRD that used `ValidationErr` type and distinct reason fields, which would imply a non-unique type in the slice :-/

thockin commented:
It makes patch very unhappy

On Thu, Jul 2, 2020 at 4:14 PM James Peach <notifications@github.com> wrote:
>
> oh, yeah. I thought that was understood :) Maybe I have just bumped my head on non-unique key fields too many times.
> …
>
> I was proposing for a CRD that used ValidationErr type and distinct reason fields, which would imply a non-unique type in the slice :-/
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

evankanderson commented:
> While we are at it, TIL that condition types are expected to be [unique](https://github.com/kubernetes/kubernetes/pull/92519).
> The current API conventions don't address that, so I had assumed that using the same condition type with a different reason would be OK. When I think about conditions as error conditions, this seems overly restrictive to me, but when I think about them as state or phase markers, uniqueness seems desirable.

I think I documented this in the new text:

> Therefore, conditions are
> represented using a list/slice, where all have similar structure.	represented using a list/slice of objects, where each condition has a similar
> structure. This collection should be treated as a map with a key of `type`.

dims commented:
@thockin @liggitt looks like @evankanderson already took care of the last round of comments a few weeks ago. Can you please peek again?

evankanderson commented:
@thockin @liggitt -- any further thoughts? Does making Conditions consumable by machines as well as humans (and by UIs, which is machines interpreting the API for humans).

lavalamp commented on contributors/devel/sig-architecture/api-conventions.md:
```diff
@@ -308,17 +308,67 @@ response reduces the complexity of these clients.
+  cause double-negative confusion.
+
+* Condition type names should describe the current observed state of the
+  resource, rather than describing the current state transitions. This
```
Sufficiently slow transitions are best thought of as states in and of themselves, so I'm not convinced about this paragraph.

lavalamp replied:
I'm ready to sign off on the other changes, I'm on the fence about whether to ask for a change here or not.

evankanderson replied:
I can update to mention slow transitions as states (LetsEncrypt cert provisioning strikes me as an example)

liggitt replied:
Yeah, this paragraph was the only one I bumped on.

I'm thinking of conditions like `Resizing=True` the PV controller or node has contemplated adding in the past, and trying to figure out what the alternative would be if this was the guidance. `NotResizing=True`, etc, to conform to this paragraph seems of the "up with which I will not put" variety.

evankanderson replied:
`ResizeInProgress`? That also provides an indication that "yes, I picked up this work, and I'm working on it now".

liggitt replied:
I guess? I don't really see a distinction between `Resizing=True` and `ResizeInProgress=True` and would probably have picked the former.

evankanderson replied:
Actually, it seems that `ResizeSucceeded` would be a better condition:

* `True` = object is at the correct size
* `Unknown` = I'm still working
* `False` = I'm sorry, Dave. I'm afraid I can't do that.

lavalamp replied:
`Resizing` is fine (and abides by the earlier statement to prefer brevity).

I think `AboutToStartResizing` and `AboutToFinishResizing` are ones we'd want to recommend against.

```
State transitions have transitions
In between to change 'em
Those have yet smaller changes too,
and so, ad infinitum
```

([original](https://en.wikipedia.org/wiki/Siphonaptera_%28poem%29))

evankanderson replied:
Added a new paragraph suggesting `Provisoned` or `SizeAllocated` as names for these "takes a while" conditions. Another good example might be certificate signing via the ACME protocol with LetsEncrypt, which can take 5 minutes or so. There's a distinction between "still working" and "it failed due to rate limit" that seems like it could be usefully captured within a single condition.

lavalamp commented on contributors/devel/sig-architecture/api-conventions.md:
```diff
@@ -308,17 +308,75 @@ response reduces the complexity of these clients.
+  ("Deploying"). Intermediate states may be indicated by setting the status of
+  the condition to `Unknown`.
+
+  * For state transitions which take a long period of time, use a condition
+    which indicates that the transition succeeded ("Provisioned" or
+    "SizeAllocated") with an intermediate state of `Unknown`. Note that this
+    pattern makes an explicit distinction between "condition not set" (which
+    means the reconciler hasn't seen the latest state) and "condition set to
+    Unknown" (which means that the reconciler has seen the state but hasn't
+    finished the reconciliation).
```
I guess your goal is that some watcher can observe progress and know that it's done? You should document how long after the operation the condition should be left there.

But I'm not convinced this is the way to do it, what if there are two resizes back to back? I don't think you can really have conditions 1:1 with operations.

I think "Resizing=true" that is present during the resize and removed after the resize is also a legitimate--and possibly less confusing--way to represent this. Seeing "ResizingSucceeded=Unknown" is disconcerting, even if it's technically accurate. 

Maybe the way to strengthen your point is to talk about how the failure of the operation will be represented. I think you can't really do that with a "Reszing" condition, but if you have another way of representing failure, it's OK. If you need a way to represent failure, then "ResizingSucceeded=false" works better.

evankanderson replied:
Actually, my concern is that the watcher can determine whether progress is being made in a generic way. (Particularly in the presence of CRDs and a variety of types, detecting "controller is stuck" vs "backend system is slow" can be quite difficult.)

I'd suggest not removing the condition when the operation is done (hence, `Provisioned` or `SizeAllocated`). The goal is twofold:

1. Presence of the condition along with `LastTransitionTime` helps to debug whether the controller has seen the most recent `spec`.

1. Ability to signal success and failure of these operations in a generic way that will stick with the resource seems useful for the "hey, could you help debug this with me" case -- a resize failed condition that disappears after 2 hours might be gone by the time an expert takes a look at the correct resource in the cluster.


As for the human ergonomics, I was hoping that the `Reason` and `Message` fields would suffice:

```yaml
status:
  conditions:
  - name: Resizing
    status: True
    Reason: FileSystemExtensionInProgress
    Message: The system is currently extending the filesystem on the disk after resize.
```

```yaml
status:
  conditions:
  - name: SizeAvailable
    status: Unknown
    Reason: FileSystemExtensionInProgress
    Message: The system is currently extending the filesystem on the disk after resize.
```

(Note that I changed this to `SizeAvailable` because there's probably a distinction between allocation failures and filesystem extension failures.)

lavalamp replied:
Maybe a way to say this is something like "conditions should not be transient, they should more or less always be on the object with the appropriate value".

But this doesn't rule out something like "Resizing", you can just set it False when the object is the correct size. At first it seems like people would need to look at the Reason to know if it's a "good" false (no resize needed) or a "bad" false (resize failed), but in the latter case we can just leave it in "resizing" (Reason is e.g. "stuck", message is "resize operation can't complete due to error foo").

evankanderson replied:
That works for me -- my concern was more specifically that the set of signalled Conditions not be transient, which this addresses. PTAL!

evankanderson commented:
As a followup to this PR, I notice that [Eric's analysis](https://github.com/kubernetes/community/pull/4521#discussion_r445882609) highlights a number of cases where developers are prefixing their Condition types, ala [`AppDBInstanceReady`](https://github.com/danisla/appdb-operator/blob/65ad0639e11cba6b394e094aca59c99759d22b89/pkg/types/appdb.go#L41) for a resource named `AppDBOperator`. Are we tired of swinging at this, or would a subsequent PR along the lines of [smarterclayton's comment](https://github.com/kubernetes/community/pull/4521#discussion_r443665374) be a good follow-on?

> Conditions summarize characteristics of object lifecycle that must be generic across multiple types, have a need to convey both state and human message, and can be extended by third parties to add more nuance to existing lifecycle.
>
> An object in Kube is a distributed state machine. The state specified to that object should always be fields. Some objects have consistent fields that describe status of the state machine. One of those is observedGeneration. Another one is conditions. Some conditions define generic, multi-object state machine characteristics like Available, Progressing, Ready.

(I don't want to derail this PR with extra load at this point, but it's been bugging me since I came back to this.)

lavalamp commented on contributors/devel/sig-architecture/api-conventions.md:
```diff
@@ -371,12 +424,12 @@ we define comprehensive state machines for objects, nor behaviors associated
-could be `Failed`. A `True` status for `Failed` would imply failure with no
-retry. An object that was still active would generally not have a `Failed`
-condition.
+An example of an oscillating condition type is `Ready`, which indicates the
```
`Resizing` is another natural example, but no need to add it here. :)

lavalamp commented:
/lgtm
/approve

Let's get this in before anyone thinks of another thing to change. There can always be a follow-up!

lavalamp commented:
> Are we tired of swinging at this, or would a subsequent PR along the lines of smarterclayton's comment be a good follow-on?

I bet we can do a followup in less time than this PR took :)

thockin reviewed (commented):
I'm still super torn on this.  This PR has been reduced almost beyond arguability, but still...

_I don't think we have a clear rule_ for when some datum wants to be a field or a condition.  Half of the examples I see here would, IMO, be better as fields.  So why are we suggesting conditions for them?

lavalamp commented:
I agree with that, but this PR has been pending for more than half a year and is an improvement over the existing text.

dims commented:
@thockin if you want to throw some outline of what you want to see, i can try to convince @evankanderson to do a follow up :)

evankanderson commented:
I'd be happy to schedule a chat with whomever is interested to try to work out some better / additional guidance.

I can say that we've found the structure of Conditions (short code plus human-readable text and the tri-state Status) very useful for communicating to humans, so I'd like to preserve that.

dims commented:
@evankanderson let's line it up for next sig-arch?

thockin commented:
I mean, I get that field-level metadata is nice, but if that's the case, we
should have it on all fields.  We shouldn't force things to fit into a
string-bool just to get metadata.  Or else the metadata is "nice" and not
"critical"

On Fri, Sep 25, 2020 at 2:59 PM Evan Anderson <notifications@github.com>
wrote:

> I'd be happy to schedule a chat with whomever is interested to try to work
> out some better / additional guidance.
>
> I can say that we've found the structure of Conditions (short code plus
> human-readable text and the tri-state Status) very useful for communicating
> to humans, so I'd like to preserve that.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/4521#issuecomment-699176010>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVHATNPK5GUGA3GZAOLSHUHF3ANCNFSM4KVRDSEQ>
> .
>

lavalamp commented:
Practical reasons not to apply this to all fields: that go doesn't do templates (yet), and objects would get very large.

One practical reason to keep it a bool is that it's simple to express what you want to watch for with kubectl tools.

My current guidance would be, conditions are good for:
* Top-level facts about the object,
* Which can change over the lifetime of the object (i.e., could be a valid target to wait for),
* And are not overly detailed (i.e., don't make a separate condition for the version of every image cached on the node),
* And have a unique owner (i.e., each condition has a single setter, but it's ok to have different setters for different conditions).

I think the "too detailed" and "valid target to wait for" tests are subjective and require some taste.

I think there's a bit of an open question when multiple instances of the same binary want to vote/set a condition; apiserver does this in multiple places and I'm not sure we're super happy about it. @deads2k
